### PR TITLE
Prevent double cloning of git repos when fetching them in parallel

### LINKED ISSUE #1206: "dest already exists" error
I have this repo https://github.com/flip111/purescript-barlow-lens I'm getting the following error

```
» spago test
Reading Spago workspace configuration...

✅ Selecting package to build: barlow-lens

Cloning https://github.com/purescript/purescript-arrays.git
file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:61132
                  throw util.fromLeft(step3);
                  ^

Error: dest already exists.
    at doRename (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:10603:15)
    at moveSync4 (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:10588:14)
    at file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:94011:56
    at runSync (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60789:21)
    at run16 (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60982:25)
    at file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60998:23
    at drain (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60816:9)
    at Object.enqueue (file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60834:11)
    at file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:60992:31
    at file:///home/flip111/.nvm/versions/node/v21.7.1/lib/node_modules/spago/bin/bundle.js:90717:30

Node.js v21.7.1
```

### PR DESCRIPTION
**Pull request by f-f**
Fix #1206: in #1275 we introduced a cache for git repos so we would not clone them more than once - however this didn't cover the case where we have a valid lockfile and we start with an empty `.spago` folder (such as the case of a fresh clone of a repo).

This patch introduces a lock for git repos so we clone them only once even when multiple fibers are trying to clone them in parallel. We also add a test that reproduces the issue exactly, and behaves well with this patch.

### CODE REVIEW

**Comment by f-f** [comment]
@fsoikin any idea what's up with windows here? The two fixtures are identical...

**Comment by fsoikin** [comment]
I will take a look later (afk right now), but my first guess would be line breaks.

File: src/Spago/Command/Fetch.purs [hunk 1]
```diff
@@ -204,11 +206,27 @@ run { packages: packagesRequestedToInstall, ensureRanges, isTest, isRepl } = do
...
+        -- and possibly cloning the same repo multiple times - so we use a lock on the git url to prevent that
+        case Map.lookup gitPackage.git gitLocks of
+          Nothing -> do
+            -- This is not supposed to happen but we just move on if it does
+            getGitPackageInLocalCache name gitPackage
+          Just lock -> do
+            -- Take the lock, do the git thing, release the lock
+            liftAff $ AVar.take lock
+            getGitPackageInLocalCache name gitPackage
+            liftAff $ AVar.put unit lock
```
Comment (fsoikin) [inline src/Spago/Command/Fetch.purs:229]:
```suggestion
        let lock = Map.lookup gitPackage.git gitLocks
        -- Take the lock, do the git thing, release the lock
        liftAff $ AVar.take `traverse_` lock
        getGitPackageInLocalCache name gitPackage
        liftAff $ AVar.put unit `traverse_` lock
```

**Comment by fsoikin** [comment]
Yep, it's the newlines. They matter sometimes, but not always. I haven't figured out precisely when, but I suspect that the Purs output has platform-dependent newlines in it. That's why it worked before, but not after you added a build, and that's why it started working after you cut out the Purs output.

I pushed a fix. I don't have a lot of time now, so I just added another `replaceAll`, but I think this whole sanitize business needs to be systematized and extracted in blocks for reuse for better ergonomics.

File: src/Spago/Command/Fetch.purs (hunk 1 above)
Comment (f-f) [reply src/Spago/Command/Fetch.purs:229]: Well spotted! It's always traverse 😄

**Comment by f-f** [comment]
Ah yeah this all makes sense. Thanks @fsoikin for pushing the fix!

**Pull request merged**